packages/react-native/ReactCommon/react/performance/timeline/PerformanceObserverRegistry.cpp

```diff
@@ -0,0 +1,53 @@
+/*
+ * Copyright (c) Meta Platforms, Inc. and affiliates.
+ *
+ * This source code is licensed under the MIT license found in the
+ * LICENSE file in the root directory of this source tree.
+ */
+
+#include "PerformanceObserverRegistry.h"
+
+namespace facebook::react {
+
+void PerformanceObserverRegistry::addObserver(const std::weak_ptr<PerformanceObserver>& observer) {
+  std::lock_guard guard{observersMutex_};
+  observers_.insert(observer);
+}
+
+void PerformanceObserverRegistry::removeObserver(const PerformanceObserver& observer) {
+  std::lock_guard guard{observersMutex_};
+  erase_if(observers_, [&](auto e) -> bool {
+    return !e.expired() && *e.lock() == observer;
+  });
+}
+
+void PerformanceObserverRegistry::removeObserver(const std::shared_ptr<PerformanceObserver>& observer) {
+  std::lock_guard guard{observersMutex_};
+  observers_.erase(observer);
+}
+
+void PerformanceObserverRegistry::emit(const facebook::react::PerformanceEntry& entry) {
```

Comment:
Nit: I'd use something like `queuePerformanceEntry` to use a similar naming as in the spec: https://w3c.github.io/performance-timeline/#queue-a-performanceentry